pkg/util/pki/parse.go

```diff
@@ -140,3 +140,161 @@ func DecodeX509CertificateRequestBytes(csrBytes []byte) (*x509.CertificateReques
+				certs = append(certs[:i], certs[i+1:]...)
+			}
+		}
+	}
+
+	var chains []*chainNode
+	// Build the nodes to the linked list
+	for i := range certs {
+		chains = append(chains, &chainNode{cert: certs[i]})
+	}
+
+	for {
+		// We have built the full chain
+		if len(chains) == 1 {
+			break
+		}
+
+		// lastChainsLength is used to ensure that at every pass, the number of
+		// tested chains gets smaller.
+		lastChainsLength := len(chains)
+		for i := 0; i < len(chains)-1; i++ {
+			for j := 1; j < len(chains); j++ {
+				if i == j {
+					continue
+				}
+
+				// attempt to add both chain together
+				chain, ok := chains[i].tryMergeChain(chains[j])
+				if ok {
+					// If adding the chains together was successful, remove inner chain from
+					// list
+					chains = append(chains[:j], chains[j+1:]...)
+				}
+
+				chains[i] = chain
+			}
+		}
+
+		// If no chains were merged in this pass, the chain can never be built
+		// flat. Error.
+		if lastChainsLength == len(chains) {
+			return nil, errors.NewInvalidData("certificate chain is malformed or broken")
+		}
+	}
+
+	// There is only a single chain left at index 0. Return chain as PEM.
+	return chains[0].toBundleAndCA()
+}
+
+// toBundleAndCA will return the PEM bundle of this chain.
+func (c *chainNode) toBundleAndCA() (*PEMBundle, error) {
+	var (
+		certs []*x509.Certificate
+		ca    *x509.Certificate
+	)
+
+	for {
+		// If the issuer is nil, we have hit the root of the chain. Assign the CA
+		// to this certificate and stop traversing.
+		if c.issuer == nil {
+			ca = c.cert
+			break
+		}
+
+		// Add this nodes certificate to the list at the end. Ready to check next
+		// node up.
+		certs = append(certs, c.cert)
+		c = c.issuer
+	}
+
+	capem, err := EncodeX509(ca)
+	if err != nil {
+		return nil, err
+	}
+
+	// If no certificates parsed, then CA is the only certificate and should be
+	// the chain
+	if len(certs) == 0 {
+		return &PEMBundle{ChainPEM: capem}, nil
+	}
+
+	// Encode cert chain
+	chainpem, err := EncodeX509Chain(certs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return chain and ca
+	return &PEMBundle{CAPEM: capem, ChainPEM: chainpem}, nil
+}
+
+// tryMergeChain will attempt to add the chain "c" and "chain" together.
+// Successful if the either chain is added to the top or bottom of the other
+// chain.
+func (c *chainNode) tryMergeChain(chain *chainNode) (*chainNode, bool) {
+	// The given chain's root has been signed by this node. Add this node on top
+	// of the given chain.
+	if chain.root().cert.CheckSignatureFrom(c.cert) == nil {
+		chain.root().issuer = c
+		return chain, true
+	}
+
+	// The given chain is the issuer of the root of this node. Add the given
+	// chain on top of the root of this node.
+	if c.root().cert.CheckSignatureFrom(chain.cert) == nil {
+		root := c.root()
+		root.issuer = chain
+		return c, true
+	}
+
+	// Chains cannot be added together.
+	return c, false
+}
+
+// Return the root most node of this chain.
```

Comment:
question: what do we mean by "root most"? Do we talk about "trust root" (from the X.509 terminology) or "tree root" (from graph theory)?

Reply:
Tree root

Reply:
I know I am nitpicking but... in the context of a tree, the root is always unique (otherwise, we would be calling it "directed acyclic graph" 😅)

Reply:
Again, not sure I understand the indention. I think "Return the root most node of this chain." is clear what it is doing and readers understand that it will be returning the root most item, both in terms of list index _and_ root most certificate.

> I suggest rewording to mention that the returned CA is the root of the tree we built, but it might not be the actual trust root since we might only have an intermediate CA.

This is already covered in the definitions above and won't be the true root of our chain as we are merging the lists. I think we're in danger of adding more comments than code complexity for an unexported self explanatory helper function.

I'll leave the comment as is.

Reply:
> in the context of a tree, the root is always unique 

Not sure I understand this comment. This root is unique for this chain?

Reply:
I think I can see what Mael is getting at; "most" feels redundant here, and "Return the root node of this chain" would be sufficient. But yeah I wouldn't rush to change it, it's no big deal.